File: maximum-xor-with-an-element-from-array.cc
#define ALL(x) (x).begin(), (x).end()
#define FOR(i, a, b) for (remove_cv<remove_reference<decltype(b)>::type>::type i = (a); i < (b); i++)
#define REP(i, n) FOR(i, 0, n)

class Solution {
public:
  vector<int> maximizeXor(vector<int>& nums, vector<vector<int>>& queries) {
    int n = nums.size(), m = queries.size();
    REP(i, m) queries[i].push_back(i);
    sort(ALL(nums));
    sort(ALL(queries), [](auto &l, auto &r) { return l[1] < r[1]; });
    vector<int> ans(m), best(m, -1), p0(n), p1(n);
    REP(i, n)
      p0[i] = i-1;
    for (int k = 1<<29; k; k >>= 1) {
      REP(i, n)
        if (p0[i] == -1)
          p1[i] = -1;
        else if (((nums[i]^nums[p0[i]]) & k) == 0)
          p1[i] = p1[p0[i]];
        else
          p1[i] = p0[i], p0[i] = p1[p0[i]];
      int i = 0;
      REP(j, m) {
        for (; i < n && nums[i] <= queries[j][1]; i++);
        int &b = best[j];
        if (b == -1)
          b = i-1;
        if (~b && ((queries[j][0]^nums[b]) & k) == 0 && ~p1[b])
          b = p1[b];
      }
    }
    REP(j, m)
      ans[queries[j][2]] = ~best[j] ? queries[j][0] ^ nums[best[j]] : -1;
    return ans;
  }
};

File: maximum-xor-with-an-element-from-array.rs
// Maximum XOR With an Element From Array
impl Solution {
    pub fn maximize_xor(mut nums: Vec<i32>, mut queries: Vec<Vec<i32>>) -> Vec<i32> {
        use std::usize::MAX;
        let n = nums.len();
        nums.sort_unstable();
        for (i, q) in queries.iter_mut().enumerate() { q.push(i as i32); }
        queries.sort_unstable_by_key(|q| q[1]);
        let mut ans = Vec::with_capacity(queries.len()); unsafe { ans.set_len(queries.len()); }
        let mut best = vec![MAX; queries.len()];
        let mut p0: Vec<usize> = (0..n).map(|i| i.wrapping_sub(1)).collect();
        let mut p1 = Vec::with_capacity(n); unsafe { p1.set_len(n); }
        for kk in (0..30).rev() {
            let k = 1 << kk;
            for i in 0..n {
                if p0[i] == MAX { p1[i] = MAX; }
                else if ((nums[i]^nums[p0[i]]) & k) == 0 { p1[i] = p1[p0[i]]; }
                else { p1[i] = p0[i]; p0[i] = p1[p0[i]]; }
            }
            let mut i = 0;
            for (q, b) in queries.iter().zip(&mut best) {
                while i < n && nums[i] <= q[1] { i += 1; }
                if *b == MAX { *b = i.wrapping_sub(1); }
                if *b != MAX && ((q[0]^nums[*b]) & k) == 0 && p1[*b] != MAX { *b = p1[*b]; }
            }
        }
        for (q, b) in queries.iter().zip(best) {
            ans[q[2] as usize] = if b == MAX { -1 } else { q[0]^nums[b] };
        }
        ans
    }
}
